xdsl/transforms/printf_to_putchar.py

```diff
@@ -0,0 +1,148 @@
+from xdsl.builder import ImplicitBuilder
+from xdsl.dialects import arith, func, scf
+from xdsl.dialects.builtin import IndexType, ModuleOp, f32, i32
+from xdsl.dialects.experimental import math
+from xdsl.dialects.printf import PrintCharOp, PrintIntOp
+from xdsl.ir import Block, MLContext
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.printer import Printer
+
+
+class LowerPrintCharToPutchar(RewritePattern):
+    """
+    Rewrite Pattern that rewrites printf.print_char to an
+    external function call
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: PrintCharOp, rewriter: PatternRewriter, /):
+        func_call = func.Call.get("putchar", [op.char], [i32])
+        # Add empty new_results, since a result is necessary for linking
+        # putchar, but the result does not exist anywhere.
+        rewriter.replace_matched_op(func_call, new_results=[])
+
+
+class ConvertPrintIntToItoa(RewritePattern):
+    """
+    Rewrite Pattern that rewrites printf.print_int to an
+    inline_itoa function
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: PrintIntOp, rewriter: PatternRewriter, /):
+        func_call = func.Call.get("inline_itoa", [op.int], [])
+        rewriter.replace_matched_op(func_call)
+
+
+def get_inline_itoa():
+    inline_itoa_func = func.FuncOp("inline_itoa", ((i32,), ()))
+    with ImplicitBuilder(inline_itoa_func.body) as (integer,):
+        zero = arith.Constant.from_int_and_width(0, i32)
+        is_negative = arith.Cmpi.get(integer, zero, "slt")
+        # If the value is zero, just print one zero
+        is_zero = arith.Cmpi.get(integer, zero, "eq")
+        is_zero_block = Block()
+        # If is_zero is true:
+        with ImplicitBuilder(is_zero_block):
+            ascii_zero = arith.Constant.from_int_and_width(48, i32)
+            # The ascii value for zero is 48 in decimal
+            PrintCharOp((ascii_zero,))
+            scf.Yield.get()
+        # If is_zero is false:
+        is_not_zero_block = Block()
+        with ImplicitBuilder(is_not_zero_block):
+            # If the value is negative, print the minus sign, return abs value
```

Comment:
The printing the minus sign and returning the absolute value could be one of those utility methods.